### DISCUSSION TITLE
K8S probe timing out when using ManagedExecutor a lot

### DISCUSSION THREAD

**Discussion by alex-martel** [quarkusio/quarkus#27007] [Q&A]
Hi all,

this possibly relates to https://github.com/quarkusio/quarkus/issues/23429 and this comment on another related issue; https://github.com/quarkusio/quarkus/issues/7893#issuecomment-616128523

We do not use much resteasy-reactive, just few custom management apis, but we do an extensive use of the ManagedExecutor. At one point kubernetes killed our application because of a timeout in the liveness probe. This happens when we run with a certain load (managed executor running async about 5k+ somewhat short tasks), called from a quartz `@Scheduled` method

we got this exception when it happened before everything died and got restarted again. Unsure if it is actually related to the actual issue since k8s killed the pod.

```
2022-07-26 18:00:48,596 ERROR [io.qua.ver.cor.run.VertxCoreRecorder] (executor-thread-39) Uncaught exception received by Vert.x: java.util.concurrent.RejectedExecutionException: event executor terminated
        at io.netty.util.concurrent.SingleThreadEventExecutor.reject(SingleThreadEventExecutor.java:923)
        at io.netty.util.concurrent.SingleThreadEventExecutor.offerTask(SingleThreadEventExecutor.java:350)
        at io.netty.util.concurrent.SingleThreadEventExecutor.addTask(SingleThreadEventExecutor.java:343)
        at io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:825)
        at io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:815)
        at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:83)
        at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:173)
        at io.vertx.core.impl.future.FutureBase.emitFailure(FutureBase.java:66)
        at io.vertx.core.impl.future.FutureImpl.tryFail(FutureImpl.java:230)
        at io.vertx.core.impl.future.PromiseImpl.tryFail(PromiseImpl.java:23)
        at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:161)
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:555)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1462)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:833)
        at com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:704)
        at com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:202)

2022-07-26 18:00:48,596 INFO  [io.quarkus] (Shutdown thread) eae stopped in 23.672s
```

It would be nice I think for that reason, to get the management apis on another port, if it could solve that particular issue. I wonder if anybody had that kind of thing, and any ways to solve it with the current quarkus implementation. I am sure I can bump the timeout, but it seems like pushing the problem, until a next time.

[a comment by an automated account is not shown]

**Comment by geoand** [comment]
Could you show an example of your Quartz usage?

**Reply by alex-martel** [reply]
Sure, I cannot put everything I have but hopefully this will give an idea. the scheduledMethod runs for less than a minute. The data list can be from 1 to 10k elements.

```java
    @Scheduled(every = "2m", delayed = "10s", concurrentExecution = SKIP)
    void scheduledMethod() {
        List<Data> data; // this comes from a multiple rest calls (multiple Unis, that we join and await on);
        CountDownLatch latch = new CountDownLatch(data.size());

        data.forEach(e -> executor.runAsync(() -> {
            try {
                // This usually does not block, but sometimes will need to do restclient (blocking, no mutiny) calls if data is 'incomplete'
                // hence why we make is run async via the ManagedExecutor.
                doSomething(d);
            } finally {
                latch.countDown();
            }
        }));

       // We make sure to stay in the call, so it does not run concurrently.
        try {
            latch.await();
        } catch (InterruptedException e) {
            log.error("Exception happened", e);
            Thread.currentThread().interrupt();
        }
    }
```

**Reply by alex-martel** [reply]
Since it seems to exhaust the managedExecutor thread pool, we are trying now with something like this; `@ManagedExecutorConfig(maxAsync = 150)`. On my laptop I can go with much higher load, and the probes always work/fast, so I am guessing there is something related to # of cpu/cores allocated on k8s as well.

**Reply by geoand** [reply]
It seems like some of the time there is a blocking REST call being made in the async code - that is a very big red flag. If you really need to make a REST call, you should make it asynchronous or else the event loop threads can very very easily all get tied up waiting for the response of these calls. Once that happens, no Vert.x IO (including the health checks) will work properly.

**Reply by alex-martel** [reply]
Thank you for your reply. Sorry for my ignorance, but I fail to see the red flag here. It is usually a common principle to put long/blocking tasks in their own thread so they don't block the main thread. In Spring for example we would create our own TaskExecutor, and make any super long task running there/blocking, without impacting the rest of the application. I have tried in the past to create my own threadpool in quarkus but it led to failures due to context issues, I think due due TCCL (example here; https://github.com/quarkusio/quarkus/issues/24896).

In the case above, what would then be my options? Would using a Uni<>, and awaiting on it cause the same result? It blocks, but the call I guess would not be on the event loop? That piece of code unfortunately was not really designed to be fully reactive.

Thank you for your time and your knowledge

**Reply by geoand** [reply]
Hi,

I might have misunderstood where your tasks are executing. I see you using `executor.runAsync`, where is this coming from?

**Reply by alex-martel** [reply]
I have this right above the scheduled method
```java
    @Inject
    ManagedExecutor executor;
```

We are now trying with something like this
```java
    @Inject
    @ManagedExecutorConfig(maxAsync = 150)
    ManagedExecutor executor;
```

**Reply by geoand** [reply]
I see. I would suggest debugging on which thread the IO is **actually** occuring.

**Reply by alex-martel** [reply]
What do you exactly mean by IO? The rest calls?

**Reply by geoand** [reply]
Yes

**Reply by alex-martel** [reply]
The rest call is running on the vertx event loop, whether I use a reactive type or not.

`DEBUG [org.jbo.res.rea.cli.log.DefaultClientLogger] (vert.x-eventloop-thread-11) Request: GET http://localhost:8091/api`
I tried changing the api to use Uni<Type>, it did run the same thread as if returning Type only.

I am unsure how to **not** make it run on the eventloop, if that's the bad thing here.

**Reply by geoand** [reply]
Do you perhaps have a sample application I can have a look at?

**Reply by alex-martel** [reply]
I did not, but now I do [here](https://github.com/manofthepeace/quarkus-restclient-eventloop)  ;). Please let me know if this is enough; It's pretty dirty and only tested with dev mode, but it shows all rest calls going through the eventloop. This seem to be causing attrition on k8s when using not enough cpu/core so not as many eventloop thread as when running with a somewhat more performant machine.

**Reply by geoand** [reply]
The client calls are being executed on the event loop and don't block, which is what you want, so from that perspective things look good.

I am now wondering if the inability to respond to health endpoint in time is due to GC pressure. What kind of limits have you applied to the k8s pods?

**Reply by alex-martel** [reply]
Thanks for the confirmation. So yes seems to be 2 things in the end because with `@ManagedExecutorConfig(maxAsync = 150)` seems like it alleviated the problem as we cannot see it anymore. So looks like exhausting the executor threads is not that good of a thing i.e. 150/200. Is there facilities to created our own threadpool in quarkus? I did not find any besides via vertx.

Also yes our pod cpu is 500m at the moment, which seem ok for most of the time, but when our scheduled task run it seems to peak and render the processing of other things not really efficient.

**Reply by geoand** [reply]
> Is there facilities to created our own threadpool in quarkus?

We don't provide anything other than the `ManagedExecutor` and the built in executors for Vert.x. But if we did, I don't think it would solve your problem.

> Also yes our pod cpu is 500m at the moment

I would advise against allocating less than 2 CPUs to JVM processes. See [this](https://developers.redhat.com/articles/2022/04/19/best-practices-java-single-core-containers) and [this](https://docs.microsoft.com/en-us/azure/developer/java/containers/kubernetes) for more information.

**Reply by alex-martel** [reply]
thank you, just to be sure that 2cpu recommendation is also for native application right?

**Reply by geoand** [reply]
Good question. @galderz any specific recommendation for ^

**Reply by galderz** [reply]
@geoand Sorry I missed this. @manofthepeace GraalVM CE only has a [serial GC](https://www.graalvm.org/22.1/reference-manual/native-image/MemoryManagement/#serial-garbage-collector) right now, so giving it 2 CPUs won't have the same effect as in JVM mode, where G1 can take advantage of the extra CPU to do some of the work concurrently. There is a G1 GC for GraalVM but that's only available in their enterprise version.
